### ISSUE TITLE
LinearLR with total_iters=0 applies start_factor permanently instead of acting as a no-op

### ISSUE THREAD

**Issue by shashvat-singham** [pytorch/pytorch#193682]
### 🐛 Describe the bug

`LinearLR` is documented to interpolate the lr factor from `start_factor` to `end_factor` across `total_iters` steps. With `total_iters=0` the interpolation window is empty, so the schedule is already complete at epoch 0 and the factor should be `end_factor` (default 1.0, i.e. a no-op). Instead, `start_factor` is applied at epoch 0 and never undone — the run trains at `start_factor * base_lr` (default 1/3) forever, silently:

```python
import torch
from torch.optim.lr_scheduler import LinearLR, PolynomialLR

opt = torch.optim.SGD([torch.nn.Parameter(torch.zeros(1))], lr=1.0)
sched = LinearLR(opt, total_iters=0)
for _ in range(3):
    opt.step(); sched.step()
    print(opt.param_groups[0]["lr"])
# 0.3333333333333333
# 0.3333333333333333
# 0.3333333333333333   <- stuck at start_factor * lr for the whole run

# contrast: PolynomialLR handles the same degenerate window correctly
opt2 = torch.optim.SGD([torch.nn.Parameter(torch.zeros(1))], lr=1.0)
sched2 = PolynomialLR(opt2, total_iters=0)
for _ in range(3):
    opt2.step(); sched2.step()
    print(opt2.param_groups[0]["lr"])
# 1.0, 1.0, 1.0
```

### Cause

In `LinearLR.get_lr`:

```python
if self.last_epoch == 0:
    return [group["lr"] * self.start_factor for group in self.optimizer.param_groups]

if self._is_initial or self.last_epoch > self.total_iters:
    return _param_groups_val_list(self.optimizer, "lr")
```

Epoch 0 applies `start_factor` unconditionally; with `total_iters=0` every subsequent epoch satisfies `last_epoch > total_iters` and freezes the lr, so the incremental interpolation that would normally walk the factor up to `end_factor` never executes. `_get_closed_form_lr` divides by `total_iters`, so the closed-form path raises `ZeroDivisionError` for the same config.

### Why this comes up

`total_iters=0` isn't something you type deliberately — it arrives programmatically, e.g. `LinearLR(opt, total_iters=warmup_steps)` where computed `warmup_steps` is 0 for a small run or a resume. The natural reading is "zero-length warmup = no warmup"; the actual behaviour is a permanent 3x lr reduction with no warning, which is easy to misattribute to the model.

### Related

This is the `LinearLR` sibling of #192655 (`ConstantLR` with `total_iters <= 0` scales permanently), with a different mechanism — there the restore branch is unreachable, here the interpolation loop is empty while epoch 0 still applies the start factor. Also adjacent to #189374 (divisor validation) and the semantics RFC #168044. Whichever way the RFC lands — validate-and-reject or treat-as-no-op — `LinearLR` needs one of the two; today it does the third, worst thing (silent permanent scaling), and `PolynomialLR` shows the no-op behaviour is already the norm elsewhere in the module.

Happy to send a PR aligned with whichever direction the maintainers prefer.

### Versions

torch 2.13.0+cpu (also verified against `main` source), Python 3.11.9, Windows.

**Comment by snehasonkusare-tech** [comment]
Hi, I'd like to work on this issue if it's not already taken.
I've traced the root cause: LinearLR.get_lr() unconditionally applies
start_factor on the first step without checking whether total_iters is
already 0, so the scaling never gets undone. The closed-form path
(_get_closed_form_lr) also divides by total_iters unconditionally, which
raises a ZeroDivisionError in the same case.
I have a fix and a couple of regression tests ready locally and will open
a PR shortly.

**Comment by shashvat-singham** [comment]
Go for it — thanks for picking it up, and your root-cause read matches mine.

One thing worth settling before you write the fix, since it decides which patch is correct: there are two defensible behaviours for `total_iters=0`, and this module currently has both.

1. **Treat as a no-op** — the schedule's window is empty, so the factor should already be `end_factor` and the lr unchanged. `PolynomialLR(total_iters=0)` already behaves this way, so there's in-repo precedent.
2. **Reject at construction** — #189374 is an open PR taking the validation route for divisor params across the schedulers.

Related context you'll probably want to reference: #192655 is the `ConstantLR` sibling of this bug (same "degenerate window leaves the factor permanently applied" outcome, reached by a different branch), #185006 covers `ConstantLR` with `factor=0`, and #168044 is the broader RFC on unifying scheduler semantics. If those land on validation-first, a no-op patch here would conflict with them.

On the `_get_closed_form_lr` `ZeroDivisionError` you mentioned: worth checking `LinearLR`, `ConstantLR` and `PolynomialLR` together, since whichever way this goes the three should agree — that's really the underlying complaint in #168044.

Happy to review the PR when it's up.

**Comment by diputs-sudo** [comment]
@snehasonkusare-tech Could you ping me when the PR is up? I'm currently looking into the broader LR scheduler semantics and related issues, so i'd like to follow the approach taken here. Thanks!

**Comment by diputs-sudo** [comment]
One nuance here: `PolynomialLR(total_iters=0)` only behaves as a no-op through the normal chainable `step()` path. Its closed-form path also divides by `total_iters`, so `PolynomialLR(total_iters=0).step(0)` raises a `ZeroDivisionError`.

So I think `LinearLR(total_iters=0)` is clearly problematic today, but it may be worth deciding the behavior at the module level rather than fixing `LinearLR` in isolation: either reject `total_iters <= 0` during construction, or consistently defines a zero length schedule as already complete. A narrow no-op fix forLinearLR would address this report while leaving the neighboring scheduler behavior inconsistent. 

Also we should consider BC here, since changing an accepted configuration from its current behavior to either a no-op or a constructor error is itself an observable behavior change. If we move forward with a fix, we should first add tests that characterize the current behavior across both the chainable and closed form paths, then decide which semantics we want to preserve or change.

**Comment by diputs-sudo** [comment]
@shashvat-singham 

One thing I'd push back on slightly. I don't think we should use #189374 as precedent for `total_iters=0` semantics while it's still open. We should settle the expected behavior and BC implications first, then make the implementations consistent with that decision, not let an implementation choice establish the semantics.

**Comment by shashvat-singham** [comment]
Agreed, and that's a fair correction — an open PR shouldn't set the semantics. I listed it as one of two possible directions rather than as settled, but I can see how the framing read as "follow #189374", and your ordering is the right one: decide the behaviour and the BC question first, then make the implementations match.

The one piece of evidence I'd still put on the table is `PolynomialLR`, since that's *shipped* behaviour rather than a proposal: `PolynomialLR(opt, total_iters=0)` is already a no-op today, and `LinearLR(opt, total_iters=0)` silently applies `start_factor` forever. So the inconsistency exists in released code independently of #189374, and whichever way you settle it, one of those two changes.

For the BC angle, the two options aren't symmetric: making `total_iters=0` a no-op silently changes the lr for anyone currently (unknowingly) getting the permanent `start_factor`, while rejecting it at construction turns that into a loud error. The second seems easier to reason about for users, but I don't have a stake — happy to follow whatever you and the maintainers land on.

**Comment by yun520-1** [comment]
`LinearLR` with `total_iters=0` should be a no-op (start_factor == end_factor == 1.0). Instead, it applies `start_factor` permanently. This is a boundary-condition bug: the scheduler never enters the interpolation loop, so it should use the end state, not the start state.

**Comment by diputs-sudo** [comment]
My main concern here is BC. `LRScheduler` has been around for a long time and there’s a lot of code relying on its current behavior, including cases we probably don’t see directly in PyTorch. I don’t think “no-op feels like the more natural behavior” is enough on its own to change that.

Both options here change existing behavior: making `total_iters=0` a no-op changes the LR for code that currently accepts it, while rejecting it means that same code starts raising. Before we pick either one, I’d like to characterize the known BC impact and assess whether that change is acceptable.

So I’m less concerned right now with whether `0` should mean no-op or error. I’d first like to understand the affected surface. In particular, I’d like to see tests covering the relevant execution paths and compositions rather than assuming this change is isolated to `LinearLR`. If we can characterize what changes, what remains unchanged, and where the behavior is currently relied upon, then we’ll have a much better basis for deciding whether to change the semantics at all, and if so, which behavior we want.
